### ISSUE TITLE
[xml] XPath 2.0 matching incorrect nodes based on boolean condition

### ISSUE THREAD

**Issue by hudec117** [pmd/pmd#2494]
**Affects PMD Version:**

6.23.0

**Description:**

Either my XPath 2.0 expression is incorrect or the XPath boolean filter is not working as expected. When attempting to match only the `userPermissions` elements where `enabled` is true, it matches regardless of the value in `enabled`.

**Code Sample demonstrating the issue:**

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Profile xmlns="http://soap.sforce.com/2006/04/metadata">
    <userPermissions>
        <enabled>true</enabled>
        <name>ViewRoles</name>
    </userPermissions>
    <userPermissions>
        <enabled>false</enabled>
        <name>ViewSetup</name>
    </userPermissions>
</Profile>
```

```
//userPermissions[
    enabled/text[boolean(@Image) = true()]
]/name/text
```

![image](https://user-images.githubusercontent.com/10981949/82313273-097a4500-99c0-11ea-8599-5439f900d962.png)

---

Although, changing the condition to text comparison works:
```
//userPermissions[
    enabled/text[@Image = 'true']
]/name/text
```
![image](https://user-images.githubusercontent.com/10981949/82314385-a7224400-99c1-11ea-8c6e-0e6e01fb30d4.png)

**Steps to reproduce:**

Please provide detailed steps for how we can reproduce the bug.

1. Start PMD Designer
2. Set language to XML
3. Copy-paste code sample from above
4. Copy-paste XPath expression and set version to 2.0
5. Observe incorrectly matched nodes

**Running PMD through:** *[Ant | Designer]*

**Comment by jsotuyod** [comment]
@hudec117 thanks for the report.

I think however, the issue arises from a wrong expectation regarding how `boolean()` performs.

As per [the official specification](https://www.w3.org/TR/xquery-operators/#func-boolean):

> * If `$arg` is the empty sequence, `fn:boolean` returns `false`.
> * If `$arg` is a sequence whose first item is a node, `fn:boolean` returns `true`.
> * If `$arg` is a singleton value of type `xs:boolean` or a derived from `xs:boolean`, `fn:boolean` returns `$arg`.
> * If `$arg` is a singleton value of type `xs:string` or a type derived from `xs:string`, `xs:anyURI` or a type derived from `xs:anyURI` or `xs:untypedAtomic`, `fn:boolean` returns `false` if the operand value has zero length; otherwise it returns `true`.
> * If `$arg` is a singleton value of any numeric type or a type derived from a numeric type, `fn:boolean` returns `false` if the operand value is `NaN` or is numerically equal to zero; otherwise it returns `true`.
> * In all other cases, `fn:boolean` raises a type error [err:FORG0006].

In your XML, `<enabled>false</enabled>` is a text node, as text, the fourth rule applies, it simply checks if it's 0 length, which it's not, and therefore, evaluates to `true`. Hence, `boolean('false') = true()` is `true`.

As this is text, evaluating against text is the correct approach. Alternatively, the very same doc for `fn:boolean()` states:

> Note:
> The result of this function is not necessarily the same as " $arg cast as xs:boolean ". For example, `fn:boolean("false")` returns the value `true` whereas `"false" cast as xs:boolean` returns `false`.

There is nothing wrong with how PMD is behaving here.
